Voter tutorial update to use stargate (#588)
cosmos/sdk-tutorials: voter-legacy/index.md

We'll be mostly interested in `vue/src/views` directory, which contains page templates of our app, `vue/src/store/index.js` handles sending transactions and receiving data from our blockchain and [`@tendermint/vue`](https://github.com/tendermint/vue/) directory, which contains components, like buttons and forms.

Inside `vue/src/store/index.js` we import [@tendermint/vue/src/store/cosmos](https://github.com/tendermint/vue/blob/develop/src/store/cosmos.js) which uses [CosmJS](https://github.com/cosmwasm/cosmjs), a library for handling wallets, creating, signing and broadcasting transactions and define a Vuex store. We'll use `entitySubmit` function for sending data to our blockchain (like a JSON representing a newly created poll), `entityFetch` for requesting a list of polls and `accountUpdate` to fetch information about our token balance.

### `vue/src/views/Index.vue`

Since we don't need the default form component replace 
```js
<sp-type-form type="poll" :fields="['title', 'options', ]" module="voter" />
```

inside of `vue/src/views/Index.vue` with a new component and a Title
 
 ```js
<SpH3>
  Polls
</SpH3>
<poll-form />
 ```

> In the `<script></script>` tags below, we import our component like this

Reviewer: ```suggestion
In the `<script></script>` tags below, import the component like this
```

```js
import * as sp from "@tendermint/vue";
import PollForm from "../components/PollForm";

export default {
  components: { PollForm, ...sp },
};
```

> For our PollForm we need to create a new directory `components` in our `vue/src/` path. In this directory we create a new file `PollForm.vue` and fill our first component with life.

Reviewer: ```suggestion
For our PollForm, you need to create a new directory `components` in your `vue/src/` path. In this directory, create a new file `PollForm.vue` and fill your first component with life.
```

Reviewer: ha! "with life" I like it

### `vue/src/components/PollForm.vue`

> We create our PollForm component. Which will have a title and two buttons.

Reviewer: ```suggestion
Create the PollForm component to have a title and two buttons.
```

```js
<template>
  <div>
    <sp-input placeholder="Title" v-model="title" />
    <div v-for="option in options">
      <sp-input placeholder="Option" v-model="option.title" />
    </div>
    <sp-button @click.native="add">+ Add option</sp-button>
    <sp-button @click.native="submit">Create poll</sp-button>
  </div>
</template>
```

in between `<script></script>` tags below this:

```js
import * as sp from "@tendermint/vue";
export default {
  components: { ...sp },
  data() {
    return {
      title: "",
      options: []
    };
  },
  methods: {
    add() {
      this.options = [...this.options, { title: "" }];
    },
    async submit() {
      const payload = {
        type: "poll",
        module: "voter",
        body: {
          title: this.title,
          options: this.options.map(o => o.title)
        }
      };
      await this.$store.dispatch("cosmos/entitySubmit", payload);
        await this.$store.dispatch("cosmos/entityFetch", payload);
    }
  }
};
```

> We also need to setup our Vue store.

Reviewer: ```suggestion
Now set up your Vue store.
```

### `vue/src/store/index.js`

```js
import Vue from "vue";
import Vuex from "vuex";
import cosmos from "@tendermint/vue/src/store/cosmos.js";

Vue.use(Vuex);

export default new Vuex.Store({
  modules: { cosmos },
});

```

> In our main `App.vue` file, we make sure to initialize the Cosmos Wallet functions that we created as follows:

Reviewer: ```suggestion
In the main `App.vue` file, initialize the Cosmos Wallet functions that you created:
```

### `vue/src/App.vue`

> In the `<script></script>` tag at the end of the file, we dispatch to initialize our own app and the cosmos Vue framework

Reviewer: ```suggestion
In the `<script></script>` tag at the end of the file, dispatch to initialize your own app and the cosmos Vue framework
```

Reviewer: If it matters, I don't see the 

> `<script></script>` tag

```js
export default {
  created() {
    this.$store.dispatch("cosmos/init");
    this.$store.dispatch("cosmos/entityFetch", {type: "poll", module: "voter"});
  },
};
```

Refresh the page, sign in with a password and create a new poll. It takes a couple of seconds to process a transaction. Now, if you visit [http://localhost:1317/voter/poll](http://localhost:1317/voter/poll) you should see a list of polls (this endpoint is defined in `x/voter/rest/queryPoll.go`):

```json
{
  "height": "0",
  "result": [
    {
      "creator": "cosmos19qqa7j73735w4pcx9mkkaxr00af7p432n62tv6",
      "id": "826477ab-0005-4e68-8031-19758d331681",
      "title": "A poll title",
      "options": ["First option", "The second option"]
    }
  ]
}
```

## Adding votes

A vote type contains poll ID and a value (string representation of the selected option).

```json
starport type vote pollID value
```

### `vue/src/views/Index.vue`

> Delete the just bootstrapped for us `<sp-type-form type="vote" :fields="['pollID', 'value', ]" module="voter" />`.

Reviewer: ```suggestion
Delete the just bootstrapped for you `<sp-type-form type="vote" :fields="['pollID', 'value', ]" module="voter" />`.
```

```
in between `<script></script>` tags below this:

```js
import * as sp from "@tendermint/vue";
import AppRadioItem from "./AppRadioItem";
import AppText from "./AppText";
import {countBy } from "lodash"
export default {
  components: { AppText, AppRadioItem, ...sp },
  data() {
    return {
      selected: ""
    };
  },
  computed: {
    polls() {
      return this.$store.state.cosmos.data["voter/poll"] || [];
    },
    votes() {
      return this.$store.state.cosmos.data["voter/vote"] || [];
    }
  },
  methods: {
    results(id) {
      const results = this.votes.filter(v => v.pollID === id);
      return countBy(results, "value");
    },
    async submit(pollID, value) {
      const type = { type: "vote" };
      const body = { pollID, value };
      await this.$store.dispatch("cosmos/entitySubmit", { ...type, module:"voter", body });
      await this.$store.dispatch("cosmos/entityFetch", {...type, module: "voter"});
    }
  }
};
```

The `PollList` component lists for every poll, all the options for that poll, as buttons. Selecting an option triggers a `submit` method that broadcasts a transaction with a "create vote" message and fetches data back from our application.

> Two components are still missing from our App, to make it a bit better looking. Let's add `AppRadioItem.vue` and `AppText.vue`.

Reviewer: ```suggestion
Two components are still missing from our app to make it a bit better looking. Add  the `AppRadioItem.vue` and `AppText.vue` components.
```

  background: rgba(0, 0, 0, 0.01);
  width: 100%;
  font-family: inherit;
  text-align: left;
  font-size: 1rem;
  font-weight: inherit;
}
.button:hover {
  box-shadow: inset 0 0 0 1px rgba(0, 0, 0, 0.1),
    0 1px 5px -1px rgba(0, 0, 0, 0.1);
}
.button:focus {
  box-shadow: inset 0 0 0 1px rgba(0, 0, 0, 0.2),
    0 1px 5px -1px rgba(0, 0, 0, 0.1);
}
.button:active {
  box-shadow: inset 0 0 0 1px rgba(0, 0, 0, 0.07);
  color: rgba(0, 0, 0, 0.7);
  transform: scale(0.998);
}
</style>
```

In the `<script></script>` tag below this:

```js
export default {
  props: {
    value: "",
  },
};
```

### `vue/src/components/AppText.vue`

```js
<template>
  <div>
    <div :class="[`${type}`]">
      <slot />
    </div>
  </div>
</template>

<style scoped>
.h1 {
  font-size: 2rem;
  font-weight: 800;
  letter-spacing: 0.03em;
  margin-bottom: 2rem;
}
.h2 {
  font-weight: 800;
  text-transform: uppercase;
  letter-spacing: 0.05em;
  margin-bottom: 0.5rem;
  margin-top: 2rem;
}
.subtitle {
  color: rgba(0, 0, 0, 0.5);
  font-size: 0.85rem;
}
</style>
```

In between `<script></script>` tags below this

```js
export default {
  props: {
    type: {
      default: "p1",
    },
  },
};
```

> Now in our `App.vue` we need to update to fetch our votes. In the `<script></script>` tag we should have the following resulting code:

Reviewer: ```suggestion
In `App.vue` you need to update to fetch votes. In the `<script></script>` tag is the resulting code:
```

### `vue/src/App.vue`

```js
export default {
  created() {
    this.$store.dispatch("cosmos/init");
    this.$store.dispatch("cosmos/entityFetch", {type: "poll", module: "voter"});
    this.$store.dispatch("cosmos/entityFetch", {type: "vote", module: "voter"});
  },
};
```

> By now should be able to see the same UI as in the first screenshot. Try creating polls and casting votes. You may notice that it's possible to cast multiple votes for one poll. This is not what we want, so let's fix this behaviour.

Reviewer: ```suggestion
By now, you should see the same UI as shown in the first screenshot. Try creating polls and casting votes. You may notice that it's possible to cast multiple votes for one poll. This behavior is not what we want, so let's fix the app.
```

## Casting votes only once

> To fix this issue we first have to understand how data is stored in our application.
> 
> We can think of our data storage as a lexicographically ordered key value store. You can loop through the entries, filter by key prefix, add, update and delete entries. It is easier to visualize the store as JSON:

Reviewer: ```suggestion
To fix this issue, you must understand how data is stored in your application.

You can think of our data storage as a lexicographically ordered key-value store. You can loop through the entries, filter by key prefix, add, update and delete entries. It is easier to visualize the store as JSON:
```

```json
{
  "poll-1a266241-c58d-4cbc-bacf-aaf939c95de1": {
    "creator": "cosmos15c6g4v5yvq0hy3kjllyr9ytlx45r936y0m6dm6",
    "id": "1a266241-c58d-4cbc-bacf-aaf939c95de1",
    "title": "Soft drinks",
    "options": ["Coca-Cola", "Pepsi"]
  },
  "vote-cd63b110-2959-45b0-8ce3-afa2fb7a5652": {
    "creator": "cosmos15c6g4v5yvq0hy3kjllyr9ytlx45r936y0m6dm6",
    "id": "cd63b110-2959-45b0-8ce3-afa2fb7a5652",
    "pollID": "1a266241-c58d-4cbc-bacf-aaf939c95de1",
    "value": "Pepsi"
  }
}
```

Both `poll-` and `vote-` are prefixes. They are added to keys for ease of filtering. By convention, prefixes are defined in `x/voter/types/key.go`.

### `x/voter/keeper/vote.go`

Whenever a user casts a vote, a new "create vote" message is handled by a handler and is passed to a keeper. Keeper takes a `vote-` prefix, adds a UUID (unique to every message) and uses this string as a key. `x/voter/keeper/vote.go`:

```go
key := []byte(types.VotePrefix + vote.ID)
```

> These strings are unique and we get duplicate votes. To fix that we need to make sure a keeper records every vote only once by choosing the right key. In our case, we can use poll ID and creator address to make sure that one user can cast only one vote per poll.

Reviewer: ```suggestion
These strings are unique but it is possible to get duplicate votes. To fix that scenario, make sure a keeper records every vote only once by choosing the right key. In this case, you can use poll ID and creator address to make sure that one user can cast only one vote per poll.
```

```go
key := []byte(types.VotePrefix + vote.PollID + "-" + string(vote.Creator))
```

Restart the application and try voting multiple times on a single poll, you'll see you can vote as many times as you want but only your most recent vote is counted.

## Introducing a fee for creating polls

> Let's make it so that creating a poll costs 200 tokens.
> 
> This feature is very easy to add. We already require users to have accounts registered, and each user has tokens on balance. The only thing we need to do is to send coins from user's account to a module account before we create a poll.

Reviewer: ```suggestion
Now make it so that creating a poll costs 200 tokens.

This feature is very easy to add. You already require users to have accounts registered and each user has tokens on balance. The only thing you need to do is to send coins from the user's account to a module account before we create a poll.
```

## `x/voter/handlerMsgCreatePoll.go`:

```go
package voter

import (
	sdk "github.com/cosmos/cosmos-sdk/types"

	"github.com/alice/voter/x/voter/keeper"
	"github.com/alice/voter/x/voter/types"
	"github.com/tendermint/tendermint/crypto"
)

func handleMsgCreatePoll(ctx sdk.Context, k keeper.Keeper, msg types.MsgCreatePoll) (*sdk.Result, error) {
	var poll = types.Poll{
		Creator: msg.Creator,
		Title:   msg.Title,
		Options: msg.Options,
	}
	moduleAcct := sdk.AccAddress(crypto.AddressHash([]byte(types.ModuleName)))
	payment, _ := sdk.ParseCoins("200token")
	if err := k.CoinKeeper.SendCoins(ctx, poll.Creator, moduleAcct, payment); err != nil {
		return nil, err
	}
	k.CreatePoll(ctx, msg)

	return &sdk.Result{Events: ctx.EventManager().Events()}, nil
}
```

> The fee payment happens before `k.CreatePoll(ctx, poll)`. This way, if a user does not have enough tokens, the application will raise an error and will not proceed to creating a poll. Make sure to have `"github.com/tendermint/tendermint/crypto"` added to the import statement (if your text editor didn't do that for you).
> 
> Now, restart the app and try creating several polls to see how this affects your token balance.
> 
> Congratulations, you have built a blockchain voting application.

Reviewer: ```suggestion
The fee payment happens before `k.CreatePoll(ctx, poll)`. This way, if a user does not have enough tokens, the application raises an error and does not proceed to creating a poll. Make sure to have `"github.com/tendermint/tendermint/crypto"` added to the import statement (if your text editor didn't do that for you).

Now, restart the app and try creating several polls to see how this affects your token balance.

Congratulations! You successfully built a blockchain voting application.
```
